core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/MakeAsyncHttpRequestStage.java

```diff
@@ -156,15 +156,26 @@ private CompletableFuture<Response<OutputT>> executeHttpRequest(SdkHttpFullReque
             }
         });
 
-        // Offload the completion of the future returned from this stage onto
-        // the future completion executor
-        responseHandlerFuture.whenCompleteAsync((r, t) -> {
-            if (t == null) {
-                responseFuture.complete(r);
-            } else {
-                responseFuture.completeExceptionally(t);
+        // Attempt to offload the completion of the future returned from this
+        // stage onto the future completion executor
+        CompletableFuture<Response<OutputT>> asyncComplete =
+            responseHandlerFuture.whenCompleteAsync((r, t) -> completeResponseFuture(responseFuture, r, t),
+                                                    futureCompletionExecutor);
+
+        // It's possible the async execution above fails. If so, log a warning,
+        // and just complete it synchronously.
+        asyncComplete.whenComplete((ignored, asyncCompleteError) -> {
+            if (asyncCompleteError != null) {
+                log.debug(() -> String.format("Could not complete the service call future on the provided "
+                                              + "FUTURE_COMPLETION_EXECUTOR. The future will be completed synchronously by thread"
+                                              + " %s. This may be an indication that the executor is being overwhelmed by too"
+                                              + " many requests, and it may negatively affect performance. Consider changing "
+                                              + "the configuration of the executor to accommodate the load through the client.",
+                                  Thread.currentThread().getName()),
+                          asyncCompleteError);
+                responseHandlerFuture.whenComplete((r, t) -> completeResponseFuture(responseFuture, r, t));
```

Comment:
Would this potentially block the event loop threads? Should we consider failing the request?

Reply:
It will be the thread that called `whenCompleteAsync`.

I don't like the idea of failing the request because that seems like a waste. At this point the response handler already completed so we have a result that we would just throw away.

Reply:
My concern is that this may not fix the issue but just delay it. It'd likely be manifested as connection acquire timeout errors, which would be a lot harder to troubleshoot, but I see your point. Playing devil's advocate, I wonder if we should change the queue to be unbounded.

Reply:
> Playing devil's advocate, I wonder if we should change the queue to be unbounded. 

This may be fine too, but could lead to increased memory consumption. Since this is a publicly accessible option, we need to guard against it in any case.

Reply:
Yeah, it's like pick your poison type of thing. I'd vote for whichever is easier to troubleshoot in the failure case.

Reply:
Closing the loop: discussed with @millems as well, we're all in agreement that completing on the I/O thread is okay, especially since completing it exceptionally would also have to happen on the I/O thread.